File: services/blockchainService.js
const { Web3 } = require('web3');

// Simple NFT contract ABI (ERC-721 standard)
const NFT_ABI = [
  {
    "inputs": [
      {"name": "to", "type": "address"},
      {"name": "tokenURI", "type": "string"}
    ],
    "name": "mint",
    "outputs": [{"name": "", "type": "uint256"}],
    "type": "function"
  },
  {
    "inputs": [{"name": "tokenId", "type": "uint256"}],
    "name": "tokenURI",
    "outputs": [{"name": "", "type": "string"}],
    "type": "function"
  }
];

async function mintNFT(craftData) {
  try {
    // Initialize Web3 (using Sepolia testnet for demo)
    const web3 = new Web3(`https://sepolia.infura.io/v3/YOUR_INFURA_KEY`);
    
    // For demo purposes, we'll simulate the minting process
    // In production, you'd need:
    // 1. Deploy an actual NFT contract
    // 2. Set up proper wallet integration
    // 3. Handle gas fees and transaction signing
    
    const mockNFTData = {
      tokenId: Math.floor(Math.random() * 1000000),
      contractAddress: process.env.CONTRACT_ADDRESS || '0x1234567890123456789012345678901234567890',
      transactionHash: '0x' + Array(64).fill(0).map(() => Math.floor(Math.random() * 16).toString(16)).join(''),
      blockNumber: Math.floor(Math.random() * 1000000) + 18000000,
      network: 'sepolia',
      metadata: {
        name: craftData.story.metadata.title,
        description: craftData.story.story,
        image: craftData.imagePath,
        attributes: [
          {
            trait_type: "Artisan",
            value: craftData.artisanName
          },
          {
            trait_type: "Craft Type",
            value: craftData.craftType
          },
          {
            trait_type: "Emotional Tone",
            value: craftData.story.metadata.emotionalTone
          },
          {
            trait_type: "Cultural Origin",
            value: craftData.story.metadata.culturalOrigin
          },
          {
            trait_type: "Estimated Hours",
            value: craftData.story.metadata.estimatedHours
          }
        ],
        external_url: `https://your-platform.com/craft/${craftData.id}`,
        created_at: craftData.createdAt
      },
      mintedAt: new Date().toISOString()
    };

    console.log('NFT minted (simulated):', mockNFTData.tokenId);
    
    return mockNFTData;

  } catch (error) {
    console.error('Blockchain minting error:', error);
    throw new Error(`Failed to mint NFT: ${error.message}`);
  }
}

async function verifyNFT(tokenId, contractAddress) {
  try {
    // For real blockchain verification, uncomment and configure:
    /*
    const web3 = new Web3(`https://sepolia.infura.io/v3/${process.env.INFURA_KEY}`);
    const contract = new web3.eth.Contract(NFT_ABI, contractAddress);
    
    // Check if token exists
    const tokenURI = await contract.methods.tokenURI(tokenId).call();
    const owner = await contract.methods.ownerOf(tokenId).call();
    
    return {
      exists: true,
      owner: owner,
      tokenURI: tokenURI,
      verified: true,
      contractAddress: contractAddress,
      tokenId: tokenId
    };
    */
    
    // Current simulation for demo
    return {
      exists: true,
      owner: '0x1234567890123456789012345678901234567890',
      tokenURI: `https://your-platform.com/api/metadata/${tokenId}`,
      verified: true,
      contractAddress: contractAddress,
      tokenId: tokenId,
      isSimulated: true
    };
  } catch (error) {
    throw new Error(`Failed to verify NFT: ${error.message}`);
  }
}

// Add function to verify transaction on blockchain
async function verifyTransaction(transactionHash, network = 'sepolia') {
  try {
    // For real verification:
    /*
    const web3 = new Web3(`https://${network}.infura.io/v3/${process.env.INFURA_KEY}`);
    const receipt = await web3.eth.getTransactionReceipt(transactionHash);
    
    if (!receipt) {
      return { verified: false, message: 'Transaction not found' };
    }
    
    return {
      verified: true,
      blockNumber: receipt.blockNumber,
      gasUsed: receipt.gasUsed,
      status: receipt.status === '0x1' ? 'success' : 'failed',
      contractAddress: receipt.contractAddress,
      transactionHash: transactionHash
    };
    */
    
    // Current simulation
    return {
      verified: true,
      blockNumber: Math.floor(Math.random() * 1000000) + 18000000,
      gasUsed: Math.floor(Math.random() * 100000) + 50000,
      status: 'success',
      transactionHash: transactionHash,
      isSimulated: true
    };
  } catch (error) {
    throw new Error(`Failed to verify transaction: ${error.message}`);
  }
}

module.exports = {
  mintNFT,
  verifyNFT,
  verifyTransaction
};

File: services/blockchainService.py

import os
import random
from datetime import datetime
from web3 import Web3

# Simple NFT contract ABI (ERC-721 standard)
NFT_ABI = [
    {
        "inputs": [
            {"name": "to", "type": "address"},
            {"name": "tokenURI", "type": "string"}
        ],
        "name": "mint",
        "outputs": [{"name": "", "type": "uint256"}],
        "type": "function"
    },
    {
        "inputs": [{"name": "tokenId", "type": "uint256"}],
        "name": "tokenURI",
        "outputs": [{"name": "", "type": "string"}],
        "type": "function"
    }
]

def mint_nft(craft_data):
    try:
        # Initialize Web3 (using Sepolia testnet for demo)
        # web3 = Web3(Web3.HTTPProvider(f"https://sepolia.infura.io/v3/{os.getenv('INFURA_KEY')}"))

        # For demo purposes, we'll simulate the minting process
        mock_nft_data = {
            "tokenId": random.randint(0, 1000000),
            "contractAddress": os.getenv("CONTRACT_ADDRESS", "0x1234567890123456789012345678901234567890"),
            "transactionHash": '0x' + ''.join([random.choice('0123456789abcdef') for _ in range(64)]),
            "blockNumber": random.randint(18000000, 19000000),
            "network": "sepolia",
            "metadata": {
                "name": craft_data["story"]["metadata"]["title"],
                "description": craft_data["story"]["story"],
                "image": craft_data["imagePath"],
                "attributes": [
                    {"trait_type": "Artisan", "value": craft_data["artisanName"]},
                    {"trait_type": "Craft Type", "value": craft_data["craftType"]},
                    {"trait_type": "Emotional Tone", "value": craft_data["story"]["metadata"]["emotionalTone"]},
                    {"trait_type": "Cultural Origin", "value": craft_data["story"]["metadata"]["culturalOrigin"]},
                    {"trait_type": "Estimated Hours", "value": craft_data["story"]["metadata"]["estimatedHours"]}
                ],
                "external_url": f"https://your-platform.com/craft/{craft_data['id']}",
                "created_at": craft_data["createdAt"]
            },
            "mintedAt": datetime.utcnow().isoformat() + "Z"
        }

        print(f"NFT minted (simulated): {mock_nft_data['tokenId']}")
        return mock_nft_data

    except Exception as error:
        print(f"Blockchain minting error: {error}")
        raise Exception(f"Failed to mint NFT: {error}")

def verify_nft(token_id, contract_address):
    try:
        # Current simulation for demo
        return {
            "exists": True,
            "owner": "0x1234567890123456789012345678901234567890",
            "tokenURI": f"https://your-platform.com/api/metadata/{token_id}",
            "verified": True,
            "contractAddress": contract_address,
            "tokenId": token_id,
            "isSimulated": True
        }
    except Exception as error:
        raise Exception(f"Failed to verify NFT: {error}")

def verify_transaction(transaction_hash, network='sepolia'):
    try:
        # Current simulation
        return {
            "verified": True,
            "blockNumber": random.randint(18000000, 19000000),
            "gasUsed": random.randint(50000, 150000),
            "status": "success",
            "transactionHash": transaction_hash,
            "isSimulated": True
        }
    except Exception as error:
        raise Exception(f"Failed to verify transaction: {error}")

